### PR TITLE
fix(s3-assets): cannot publish a file without extension

### LINKED ISSUE #30471: aws-s3-assets: Asset publish fails on files with no extension if outputType is BundlingOutput.SINGLE_FILE
### Describe the bug

I'm using S3 Assets to build an executable which is downloaded on my AL2023 EC2 instances and run at startup. UNIX executables commonly have no file extension, so I named my output file `main`. Since there is only 1 output file, I used the SINGLE_FILE outputType. After bundling completes but before the asset is uploaded, `cdk deploy` will fail with the exception `EISDIR: illegal operation on a directory, read`.

### Expected Behavior

The bundled asset should have uploaded to S3 successfully. I am able to upload the asset from the cdk.out temporary directory to S3 via the console without issue.

### Current Behavior

The following is the cdk deploy output:

```
Bundling asset AssetTestStack/asset/Stage...

✨  Synthesis time: 6.15s

AssetTestStack:  start: Building 9282ebaea97fbf485dc50bbffd97ef346bb60994b53468da7becca020a3dbbb0:current_account-current_region
AssetTestStack:  success: Built 9282ebaea97fbf485dc50bbffd97ef346bb60994b53468da7becca020a3dbbb0:current_account-current_region
AssetTestStack:  start: Publishing 9282ebaea97fbf485dc50bbffd97ef346bb60994b53468da7becca020a3dbbb0:current_account-current_region
AssetTestStack:  start: Building 2ea762ee87d2dc20644f7ffe229b3f751c9831030044b4761eaef6d4688761d3:current_account-current_region
AssetTestStack:  success: Built 2ea762ee87d2dc20644f7ffe229b3f751c9831030044b4761eaef6d4688761d3:current_account-current_region
AssetTestStack:  start: Publishing 2ea762ee87d2dc20644f7ffe229b3f751c9831030044b4761eaef6d4688761d3:current_account-current_region
AssetTestStack:  fail: EISDIR: illegal operation on a directory, read
AssetTestStack:  success: Published 2ea762ee87d2dc20644f7ffe229b3f751c9831030044b4761eaef6d4688761d3:current_account-current_region

 ❌ Deployment failed: Error: Failed to publish asset 2ea762ee87d2dc20644f7ffe229b3f751c9831030044b4761eaef6d4688761d3:current_account-current_region
    at Deployments.publishSingleAsset (/usr/local/lib/node_modules/aws-cdk/lib/index.js:445:11645)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.publishAsset (/usr/local/lib/node_modules/aws-cdk/lib/index.js:445:197443)
    at async /usr/local/lib/node_modules/aws-cdk/lib/index.js:445:181344

Failed to publish asset 2ea762ee87d2dc20644f7ffe229b3f751c9831030044b4761eaef6d4688761d3:current_account-current_region
```

Interestingly, the key of the asset that failed is incorrect. `2ea762ee87d2dc20644f7ffe229b3f751c9831030044b4761eaef6d4688761d3` is the CloudFormation template JSON which uploaded to S3 successfully (as per the Published message). `9282ebaea97fbf485dc50bbffd97ef346bb60994b53468da7becca020a3dbbb0` is the asset which bundled successfully but failed to upload.

### Reproduction Steps

The following is a stack I created to isolate the issue. In this case, the asset is a Go module located at `/assets/main` but any code that compiles to a single file should produce the same result.

```
import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as s3a from 'aws-cdk-lib/aws-s3-assets'

export class AssetTestStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const asset = new s3a.Asset(this, 'asset', {
            path: './assets/main',
            bundling: {
                image: cdk.DockerImage.fromRegistry('golang:1.21.5'),
                entrypoint: ["bash", "-c"],
                command: ["GOARCH=arm64 GOOS=linux go build -o /asset-output/main ."],
                outputType: cdk.BundlingOutput.SINGLE_FILE,
            },
        });
    }
}
```

### Possible Solution

Changing the name of the output file to `main.bin` fixed the issue, but it took me a while to figure out what was causing this error since I had multiple S3 assets in the stack where I originally encountered this issue. I would assume the error comes down to the CDK incorrectly parsing a file with no extension as a directory.

### Additional Information/Context

_No response_

### CDK CLI Version

2.144.0 (build 5fb15bc)

### Framework Version

_No response_

### Node.js Version

Node.js v18.17.1

### OS

MacOS 12.6

### Language

TypeScript

### Language Version

TypeScript (5.4.5)

### Other information

_No response_

### PR DESCRIPTION
**Pull request by sakurai-ryo**
### Issue # (if applicable)

Closes #30471 

###

### Reason for this change
Publishing a file with no extension using the `Asset` class with `BundlingOutput.SINGLE_FILE` and `AssetHashType.SOURCE`(default), as shown below, will result in an error `fail: EISDIR: illegal operation on a directory, read`, and publishing will fail.
```ts
export class AssetTestStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
      super(scope, id, props);

      const asset = new s3assets.Asset(this, 'asset', {
          path: path.join(__dirname, '../assets/main'),
          bundling: {
              image: DockerImage.fromRegistry('golang:1.21'),
              entrypoint: ["bash", "-c"],
              command: ["echo 123 > /asset-output/main"], // a file without extension
              outputType: BundlingOutput.SINGLE_FILE,
          },
      });
      new CfnOutput(this, 'AssetHash', { value: asset.assetHash });
  }
}
```

This is because the path in `*.asset.json` is different from the actual file path.
The `*.asset.json` expects the file to be in `asset.bead5b2c0d128650228f146d2326d5f3cbfb36738a9383fc6a09b1e9278803f0`, but when I check the `cdk.out` directory, I see that `asset.bead5b2c0d128650228f146d2326d5f3cbfb36738a9383fc6a09b1e9278803f0` is a directory, not a file.
```json
{
  "version": "36.0.0",
  "files": {
    "bead5b2c0d128650228f146d2326d5f3cbfb36738a9383fc6a09b1e9278803f0": {
      "source": {
        "path": "asset.bead5b2c0d128650228f146d2326d5f3cbfb36738a9383fc6a09b1e9278803f0",
        "packaging": "file"
      },
      "destinations": {
        "current_account-us-east-1": {
          "bucketName": "cdk-hnb659fds-assets-${AWS::AccountId}-us-east-1",
          "objectKey": "bead5b2c0d128650228f146d2326d5f3cbfb36738a9383fc6a09b1e9278803f0.bead5b2c0d128650228f146d2326d5f3cbfb36738a9383fc6a09b1e9278803f0",
          "region": "us-east-1",
          "assumeRoleArn": "arn:${AWS::Partition}:iam::${AWS::AccountId}:role/cdk-hnb659fds-file-publishing-role-${AWS::AccountId}-us-east-1"
        }
      }
    }
}
```
```bash
cdk.out
├── SampleStack.assets.json
├── SampleStack.template.json
├── asset.bead5b2c0d128650228f146d2326d5f3cbfb36738a9383fc6a09b1e9278803f0
│   └── main
├── cdk.out
├── manifest.json
└── tree.json
```

If I change it to a file with an extension, as shown below, I see that the file with the extension is staged under `cdk.out` dir, and the asset is published successfully.
```ts
export class AssetTestStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
      super(scope, id, props);

      const asset = new s3assets.Asset(this, 'asset', {
          path: path.join(__dirname, '../assets/main'),
          bundling: {
              image: DockerImage.fromRegistry('golang:1.21'),
              entrypoint: ["bash", "-c"],
              command: ["echo 123 > /asset-output/main.bin"], // a file with an extension
              outputType: BundlingOutput.SINGLE_FILE,
          },
      });
      new CfnOutput(this, 'AssetHash', { value: asset.assetHash });
  }
}
```
```bash
cdk.out
├── SampleStack.assets.json
├── SampleStack.template.json
├── asset.dc5ce447844d7490834e46df016edc7f671b4fae19ab55b6c78973dcb5af98f8
│   └── main.bin
├── asset.dc5ce447844d7490834e46df016edc7f671b4fae19ab55b6c78973dcb5af98f8.bin # !! staged file here !!
├── cdk.out
├── manifest.json
└── tree.json
```
```json
{
  "version": "36.0.0",
  "files": {
    "dc5ce447844d7490834e46df016edc7f671b4fae19ab55b6c78973dcb5af98f8": {
      "source": {
        "path": "asset.dc5ce447844d7490834e46df016edc7f671b4fae19ab55b6c78973dcb5af98f8.bin",
        "packaging": "file"
      },
      "destinations": {
        "current_account-us-east-1": {
          "bucketName": "cdk-hnb659fds-assets-${AWS::AccountId}-us-east-1",
          "objectKey": "dc5ce447844d7490834e46df016edc7f671b4fae19ab55b6c78973dcb5af98f8.bin",
          "region": "us-east-1",
          "assumeRoleArn": "arn:${AWS::Partition}:iam::${AWS::AccountId}:role/cdk-hnb659fds-file-publishing-role-${AWS::AccountId}-us-east-1"
        }
      }
    }
}
```

Files without extensions must be staged correctly in order to be published correctly.

### Description of changes
The directory to write the bundling output is usually `cdk.out/asset.{asset hash}`.
If the extension exists, it will be renamed from `cdk.out/asset.{asset hash}/{asset file name}` to `cdk.out/{asset hash}.{asset file extension}`.
https://github.com/aws/aws-cdk/blob/c826d8faaeb310623eb9a1a1c82930b679768007/packages/aws-cdk-lib/core/lib/asset-staging.ts#L392

If the extension does not exist, the file name `cdk.out/asset.{asset hash}` (without extension) will be the same as the directory where bundling output is written.
Therefore, the file is already considered staged and will not be staged correctly.
https://github.com/aws/aws-cdk/blob/c826d8faaeb310623eb9a1a1c82930b679768007/packages/aws-cdk-lib/core/lib/asset-staging.ts#L383

Therefore, in such cases, I fix to change the file name by adding a suffix such as `noext` after the file name so that the file is correctly renamed.

### Description of how you validated changes
unit tests and integ test.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by sakurai-ryo** [comment]
@Mergifyio update

**Review by rix0rrr — COMMENTED** [review]
Thanks for the deep dive! 

I wonder if we can simplify this a bit, but apart from that looks good!

File: packages/aws-cdk-lib/core/lib/asset-staging.ts [hunk 1]
```diff
@@ -511,6 +515,15 @@ export class AssetStaging extends Construct {
         throw new Error('Unknown asset hash type.');
     }
   }
+
+  private renderStagedPath(sourcePath: string, targetPath: string): string {
+    // Add a suffix to the asset file name
+    // because when a file without extension is specified, the source directory name is the same as the staged asset file name.
+    if (this.hashType === AssetHashType.SOURCE && path.dirname(sourcePath) === targetPath) {
```
[the remark this answers is not in the document]
Comment (sakurai-ryo) [reply packages/aws-cdk-lib/core/lib/asset-staging.ts]: Changed the conditional to `this.hashType !== AssetHashType.OUTPUT` and added the reason to the comment.
https://github.com/aws/aws-cdk/pull/30597/commits/9f5adb574f11baa269764e866956df05f3c5ae8d

File: packages/aws-cdk-lib/core/lib/stack-synthesizers/asset-manifest-builder.ts [hunk 2]
```diff
@@ -24,6 +24,9 @@ export class AssetManifestBuilder {
   public defaultAddFileAsset(stack: Stack, asset: FileAssetSource, target: AssetManifestFileDestination) {
     validateFileAssetSource(asset);
 
+    // If the fileName begins with 'asset.', remove it here
+    // because when a file without extension is added to the manifest, the path.extension() will return an file hash.
+    // ex) path.extension('asset.dc5ce447844') => 'dc5ce447844'
```
[the remark this answers is not in the document]
Comment (sakurai-ryo) [reply packages/aws-cdk-lib/core/lib/stack-synthesizers/asset-manifest-builder.ts]: Deleted comment.
I forgot to delete it.
https://github.com/aws/aws-cdk/pull/30597/commits/69889ecb5b76d922fc86eee23326738db72369dd

**Comment by sakurai-ryo** [comment]
@rix0rrr 
Thanks for your review, and sorry for the late response.
 I've addressed all of your comments.

**Review by rix0rrr — APPROVED** [review]
<3

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**